src/client/address_book.cc

```diff
@@ -55,221 +55,292 @@
+    m_SubscriberUpdateTimer->async_wait(
+        std::bind(
+            &AddressBook::SubscriberUpdateTimer,
+            this,
+            std::placeholders::_1));
   }
 }
 
-bool AddressBookFilesystemStorage::GetAddress(
-    const i2p::data::IdentHash& ident,
-    i2p::data::IdentityEx& address) const {
-  auto filename = GetPath() / (ident.ToBase32() + ".b32");
-  std::ifstream f(filename.string(), std::ifstream::binary);
-  if (f.is_open()) {
-    f.seekg(0, std::ios::end);
-    size_t len = f.tellg();
-    if (len < i2p::data::DEFAULT_IDENTITY_SIZE) {
-      LogPrint(eLogError,
-          "AddressBookFilesystemStorage: file ", filename, " is too short. ", len);
-      return false;
+void AddressBook::LoadPublishers() {
+  // TODO(unassigned): this is a one-shot: we won't be able to
+  // edit publisher's file manually with any effect after router start
+  if (m_Subscribers.empty()) {
+    auto publishers = GetDefaultPublishersFilename();
+    LogPrint(eLogInfo, "AddressBook: loading publisher file ", publishers);
+    std::ifstream file(
+        i2p::util::filesystem::GetFullPath(publishers),
+        std::ofstream::in);
+    if (file.is_open()) {
+      std::string publisher;
+      while (!file.eof()) {
+        getline(file, publisher);
+        if (!publisher.length())
+          continue;  // skip empty line
+        // Perform sanity test for valid URI
+        i2p::util::http::HTTP http(publisher);
+        if (!http.GetURI().is_valid()) {
+          LogPrint(eLogWarn,
+              "AddressBook: invalid/malformed publisher address, skipping");
+          continue;
+        }
+        m_Subscribers.push_back(
+            std::make_unique<AddressBookSubscriber>(*this, publisher));
+      }
+      LogPrint(eLogInfo,
+          "AddressBook: ", m_Subscribers.size(), " publishers loaded");
+    } else {
+      LogPrint(eLogWarn,
+          "AddressBook: ", publishers, " not found; we'll use default publisher");
+      // TODO(anonimal): create default publisher file if file is missing
     }
-    f.seekg(0, std::ios::beg);
-    auto buf = std::make_unique<std::uint8_t[]>(len);
-    f.read(reinterpret_cast<char *>(buf.get()), len);
-    address.FromBuffer(buf.get(), len);
-    return true;
   } else {
-    return false;
+    LogPrint(eLogError, "AddressBook: publisher(s) already loaded");
   }
 }
 
-void AddressBookFilesystemStorage::AddAddress(
-    const i2p::data::IdentityEx& address) {
-  auto filename = GetPath() / (address.GetIdentHash().ToBase32() + ".b32");
-  std::ofstream f(filename.string(), std::ofstream::binary | std::ofstream::out);
-  if (f.is_open()) {
-    size_t len = address.GetFullLen();
-    auto buf = std::make_unique<std::uint8_t[]>(len);
-    address.ToBuffer(buf.get(), len);
-    f.write(reinterpret_cast<char *>(buf.get()), len);
+void AddressBook::SubscriberUpdateTimer(
+    const boost::system::error_code& ecode) {
+  if (ecode != boost::asio::error::operation_aborted) {
```

Comment:
[This](http://www.boost.org/doc/libs/1_61_0/doc/html/boost_asio/reference/basic_deadline_timer/async_wait.html) seems to say that `operation_aborted` is the only possible error. But we should be defensive and check for any error.